Implement constant folding in the IR backend for JVM

ting-yuan opened the pull request:
The newly added pass folds the same set of constant functions of the
current backend.

Change-Id: I739bcb3e0fc549a87123422d7ca46e44d61f24dd

dnpetrov commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    private fun buildIrConstant(call: IrCall, v: Any): IrExpression {
+        return if (call.type.isInt()) {
```
Minor (code style): replace if-else if-... with `when`.

ting-yuan replied:
Done

dnpetrov commented on compiler/testData/codegen/box/primitiveTypes/kt2269.kt:
```diff
@@ -1,13 +1,16 @@
 fun box() : String {
-    230?.toByte()?.hashCode()
-    9.hashCode()
+    // Just hard enough that the test won't get optimized away at compile time.
```
Were there any issues with this tests involving constant folding?

ting-yuan replied:
With this pass and some branch optimizations by Mads, the test simply returns OK :) The conditions are optimized away:
```
...
LDC "OK"
ARETURN
```

max-kammerer replied:
@madsager seems it's a bug in your optimizer:
```
fun box() {
    val y: String?
    if (1 == 0) { //no nop and linenumber in bytecode        
        y = "123"
    }
    else {
        y = null
    }    
}
```

madsager replied:
Thanks Mike, I agree. Ting-Yuan, we should work together on finding a way to fix that.

ting-yuan replied:
I've disabled these 2 tests and let's address it in another PR. This one is probably too large already.

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    private fun buildIrConstant(call: IrCall, v: Any): IrExpression {
+        return if (call.type.isInt()) {
```
Pls, replace with 'when'

ting-yuan replied:
Done

dnpetrov commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt:
```diff
@@ -0,0 +1,348 @@
+ *
+ * For example, the expression "O" + 'K' + (1.toLong() + 2.0) will be folded to "OK3.0" at compile time.
+ */
+class FoldConstantLowering(private val context: JvmBackendContext) : IrElementTransformerVoid(), FileLoweringPass {
```
This lowering pass adds quite some functionality to the JVM_IR back-end. I'd like to see some test for that.
Ideally, this should include two kinds of tests:
1. Box tests - to make sure everything works as expected;
2. Bytecode tests - to make sure that constant expressions are, indeed, folded.
Look into compiler/testData/codegen/box and compiler/testData/codegen/bytecodeText.

Pay attention to tests that can cause exceptions - e.g., division by zero.
```
fun test1() = 1 / 0 // ArithmeticException at run-time
fun test2() = 1.0 / 0.0 // Infinity
```

Also make sure that floating-point number comparisons work fine with respect to IEEE754 vs total order (various "special" numbers such as `-0.0` and `java.lang.Double.NaN`).

ting-yuan replied:
It looks like that most of the box tests are already there in the code base? Although only either constant folding or generated bytecode will be tested, but not both; If an expression is folded, the bytecode generated won't be tested, and vice versa.

Since this is constant folding, the result should be a constant stored in the text. I'll start by adding bytecode tests and check whether there are box tests which are folded/optimized away.

ting-yuan replied:
Please take a look at the new commit. They are added but not enabled yet; Because the current backed doesn't fold those comparisons, the new tests would fail on it. This [PR](https://github.com/JetBrains/kotlin/pull/2137) tries to fix that.

max-kammerer commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt:
```diff
@@ -0,0 +1,348 @@
+        irFile.transformChildrenVoid(object : IrElementTransformerVoid() {
+            override fun visitCall(expression: IrCall): IrExpression {
+                expression.transformChildrenVoid(this)
+                return when (expression.valueArgumentsCount) {
```
We have 2 representations for compare and equals operation in IR:
1 < 1 is binary operation with 2 argument 
1.compareTo(1) - call with dispatch receiver and 1 argument 
Maybe call.receiverAndArgs()  would be worth
Seems this cases are not tested.

Please also note that:
-0.0 < 0.0 calculation is performed in IEEE754 arithmetics and
-0.0.compareTo(0.0) in non-IEEE754 
This affects  registerSymmetricBinaryOp<Double, Double>(DOUBLE, opType, "compareTo") { a, b -> a.compareTo(b) } same applies for == and equals operation. (Similar with Floats)

Seems other operations should be equals to each other (+ and plus and so on), but maybe I miss smth @dnpetrov

ting-yuan replied:
Thanks for the pointers! I've been wondering where those IrBuiltins are used :) Please take a look at the new commit, where the support of folding some IrBuiltins, mainly for those comparisons, are added.

dnpetrov commented on compiler/testData/codegen/box/primitiveTypes/kt2269.kt:
```diff
@@ -1,13 +1,16 @@
 fun box() : String {
-    230?.toByte()?.hashCode()
-    9.hashCode()
+    // Just hard enough that the test won't get optimized away at compile time.
```
BlackBoxCodegenTestGenerated$PrimitiveTypes.testKt2269 fails: 
Unresolved reference: toInt (4,27) in /kt2269.kt

max-kammerer replied:
seems // WITH_RUNTIME directive is required

ting-yuan replied:
Updated with another approach so that  runtime isn't required. Also verified with :compiler:tests

udalov commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt:
```diff
@@ -0,0 +1,385 @@
+            registerUnaryOp<T>(operandType, "toString") { it.toString() }
+        }
+
+        init {
```
Is it possible to somehow use operations in `OperationsMapGenerated.kt` instead of manually enumerating them here? That would also make it future-proof in case we add more built-in operations in the future (like we did with `rem` in 1.1 for example)

ting-yuan replied:
Thanks for the pointer! I updated the PR quite a bit to reuse OperationsMapGenerated.kt. Please take a look again.

max-kammerer commented on compiler/testData/codegen/bytecodeText/deadCodeElimination/boxing.kt:
```diff
@@ -2,16 +2,16 @@ class A
 fun foo(x: Any?) {}
 
-fun box() {
+fun box(u: Int) {
```
Why you touch this test?

ting-yuan replied:
Similar to changes in concat.kt and primitiveToString.kt, this prevents the relevant part of the test being optimized away. Without this change, it would be:

expected:
<0 IFNULL
[1 IFNONNULL
1] throwNpe
0 ATHROW
> but was:
<0 IFNULL
[0 IFNONNULL
0] throwNpe
0 ATHROW

Namely, the check and potential throw for z!! would be optimized away.

max-kammerer replied:
@ting-yuan I would prefer to add new tests with suggested changes (for this one and below). And for kt2269.kt

max-kammerer replied:
@ting-yuan change in concat.kt is still totatly unclear for me, maybe this change is redundant after pr2132.
For primitiveToString.kt it would be best to add new test with suggested change

ting-yuan replied:
@max-kammerer The test in concat.kt has a flew IMHO. Because the constants get folded into one, there is only one concatenation after constant folding (in both backends): "12345.06.07" + A(). Because the test looks for exactly 1 StringBuilder, it does not test whether the unnecessary StringBuilders are eliminated or not.

ting-yuan replied:
@max-kammerer I touched kt2269.kt because :compiler:test fails and complains that kt2269.kt is a passing test disabled. So I have either to:
1. remove // IGNORE_BACKEND: JVM_IR so that it can pass falsely. I'll add kt2269-jvm-ir.kt
2. change the test as in PR to make it test what it meant to test.

Should I just go for 1, or other recommendations?

ting-yuan replied:
@max-kammerer I've forked those tests with proposed changes and disabled the original tests for JVM_IR, if they fail. Basically testABC.kt is forked into testABCNotOptimizable.kt.

max-kammerer commented on compiler/testData/codegen/bytecodeText/deadCodeElimination/simpleConstructorNotRedundant.kt:
```diff
@@ -1,8 +1,8 @@
 class A
-fun box() {
+fun box(u: Int) {
```
and this?

ting-yuan replied:
Similarly, the check and throw for y!! would be optimized away without this.

udalov commented on compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/FoldConstantLowering.kt:
```diff
@@ -0,0 +1,385 @@
+                registerBinaryOp<String, Any>(STRING, type, "equals") { a, b -> a.equals(b) }
+            }
+
+            // IrBuiltins
```
It would be awesome to extract names used here into constants in IrBuiltIns and use them where the corresponding operators are created, to be able to quickly see that both refer to the same concept

ting-yuan replied:
Done.

ting-yuan commented:
Hi @max-kammerer @udalov @dnpetrov,

Besides the changes to
1. tests.
2. factoring out operator names 
3. reuse OperationMapGenerated,

I also tried to get rid of descriptors. Please take a look again.

max-kammerer commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/descriptors/IrBuiltIns.kt:
```diff
@@ -215,5 +215,19 @@ class IrBuiltIns(
         val KOTLIN_INTERNAL_IR_FQN = FqName("kotlin.internal.ir")
+
+        const val OPERATOR_NAME_LESS = "less"
+        const val OPERATOR_NAME_LESS_OR_EQUAL = "lessOrEqual"
```
Maybe it's best to extract const's to separate object:

```
class IrBuiltIns {
   .....
    object OperatorNames {
        const val LESS = "less"
       ....
     }
}
```
cc @udalov

max-kammerer commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/constants/evaluate/ConstantExpressionEvaluator.kt:
```diff
@@ -1168,3 +1122,96 @@ fun CompileTimeConstant<*>.isStandaloneOnlyConstant(): Boolean {
+    return evaluateBinaryAndCheck(name, receiverType, receiverValue, parameterType, parameterValue)
+}
+
+internal fun evaluateBinaryAndCheck(
```
private

max-kammerer commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/constants/evaluate/ConstantExpressionEvaluator.kt:
```diff
@@ -1168,3 +1122,96 @@ fun CompileTimeConstant<*>.isStandaloneOnlyConstant(): Boolean {
+    return evaluateUnaryAndCheck(name, typeStrToCompileTimeType(typeStr), value)
+}
+
+internal fun evaluateUnaryAndCheck(name: String, type: CompileTimeType<*>, value: Any, tracer: () -> Unit = {}): Any? {
```
private

max-kammerer commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/constants/evaluate/ConstantExpressionEvaluator.kt:
```diff
@@ -1168,3 +1122,96 @@ fun CompileTimeConstant<*>.isStandaloneOnlyConstant(): Boolean {
+    }
+}
+
+internal fun typeStrToCompileTimeType(str: String) = when (str) {
```
private

max-kammerer commented on compiler/frontend/src/org/jetbrains/kotlin/resolve/constants/evaluate/ConstantExpressionEvaluator.kt:
```diff
@@ -1168,3 +1122,96 @@ fun CompileTimeConstant<*>.isStandaloneOnlyConstant(): Boolean {
     }
 }
+
+internal fun isZero(value: Any?): Boolean {
```
private

ting-yuan commented:
Thanks for the review! @max-kammerer. Please take a look at the updated PR again.

I'm going to resolve the comments about the tests and will remove // IGNORE_BACKEND: JVM_IR for the 2 lineNumber tests once https://github.com/JetBrains/kotlin/pull/2153 is merged.

max-kammerer commented on compiler/ir/ir.tree/src/org/jetbrains/kotlin/ir/descriptors/IrBuiltIns.kt:
```diff
@@ -251,4 +251,20 @@ class IrBuiltIns(
+    object OperatorNames {
+        const val OPERATOR_NAME_LESS = "less"
+        const val OPERATOR_NAME_LESS_OR_EQUAL = "lessOrEqual"
+        const val OPERATOR_NAME_GREATER = "greater"
```
@ting-yuan Please remove also OPERATOR_NAME prefixes (it wasn't clearly mentioned but presented in suggested example in https://github.com/JetBrains/kotlin/pull/2111/files#r264669713)

max-kammerer commented:
@ting-yuan Please rebase: #2153 is merged

ting-yuan commented:
@max-kammerer Thanks, I've rebased and removed `// IGNORE_BACKEND: JVM_IR` from the 2 line number tests, which are fixed in #2153. Also verified again with `compiler:test`.

max-kammerer commented:
@ting-yuan https://github.com/JetBrains/kotlin/pull/2111/files#r265006614 is still missed

ting-yuan commented:
@max-kammerer Sorry, I missed that. Please take a look again.

max-kammerer commented on compiler/testData/codegen/bytecodeText/constants/comparisonFalse.kt:
```diff
@@ -0,0 +1,22 @@
+// TARGET_BACKEND: JVM_IR
```
@ting-yuan Please add box tests for  comparisonFalse/comparisonTrue

ting-yuan replied:
Done!

max-kammerer commented:
@ting-yuan What is 'Change-Id ' in  commit message?

max-kammerer commented:
Please also mute JS_IR backend for newly added tests:
https://teamcity.jetbrains.com/viewLog.html?buildId=2074413&tab=buildResultsDiv&buildTypeId=Kotlin_dev_Compiler

ting-yuan commented:
@max-kammerer Thanks for catching this. I've removed `Change-Id`s and muted the 2 box tests for JR_IR.

To ensure the quality of pull request and make the review quicker, we use [Gerrit](https://www.gerritcodereview.com/) internally to have a first-round review before sending PRs. Gerrit relies on `Change-Id` to keep track of the patch history.

max-kammerer commented:
@ting-yuan Many thanks!
